torch/csrc/autograd/python_engine.cpp

```diff
@@ -201,9 +202,11 @@ PyObject *THPEngine_run_backward(THPEngine *self, PyObject *args, PyObject *kwar
       if (is_leaf) {
           grad_fn = input_var->cdata->grad_accumulator.lock();
       }
-      THPUtils_assert(grad_fn, "One of the differentiated Variables appears to not have "
-          "been used in the graph");
-      THPUtils_assert(grad_fn->is_executable, "One of the differentiated Variables does "
+      if (input_var->cdata->requires_grad) {
+        THPUtils_assert(grad_fn, "One of the differentiated Variables appears to not have "
+            "been used in the graph");
+      }
```

Comment:
What leaf Variable case are you thinking of? Here are the error messages when `input_var` is a leaf:

1) Leaf does not require grad, used in graph:
`"One of the differentiated Variables does not require grad"`
(unchanged)

2) Leaf does not require grad, unused in graph:
`"One of the differentiated Variables does not require grad"`
Previously:
`"One of the differentiated Variables appears to not have been used in the graph"`

3) Leaf requires grad and never used:
`"One of the differentiated Variables appears to not have been used in the graph"`
(unchanged)

4) Leaf requires grad, used for different operations, but not in this graph:
`"differentiated input is unreachable"`
(unchanged)

The error message for (2) seems fine to me, since input_var always needs to have `requires_grad=True`.